### ISSUE TITLE
Add "I don't support Internet Explorer" to the list of versions

### ISSUE THREAD

**Issue by colonelpopcorn** [HubSpot/youmightnotneedjquery#222]
I really enjoy using this tool to see what native APIs I can use over jQuery. However, on my side projects, I don't really ever think about supporting Internet Explorer. Could there be an option to showcase newer, more experimental features of browsers (async/await, modules, fetch API, etc.) that are much newer/more up-to-date than anything IE has?

**Comment by Richienb** [comment]
For when this feature is implemented, here are modern implementations of some solutions:

# JSON

```js
const data = await (await fetch('/my/url')).json();
```

# Post

```js
await fetch('/my/url', {
	method: 'POST',
	headers: {
		'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'
	},
	body: JSON.stringify(data)
});
```

# Request

```js
const data = await (await fetch('/my/url')).text();
```

# After

```js
target.after(element);
```

# Before

```js
target.before(element);
```

# Each

```js
document.querySelectorAll(selector).forEach((el, i) => {

});
```

# Empty

```js
el.replaceChildren();
```

# Index

```js
[...el.parentNode.children].indexOf(el);
```

# Matches Selector

```js
el.matches('.my-class');
```

# Offset

```js
const {top, left} = el.getBoundingClientRect();

{
	top: top + document.body.scrollTop,
	left: left + document.body.scrollLeft
}
```

# Position

```js
const {offsetLeft: left, offsetTop: top} = el
```

# Position Relative To Viewport

```js
const {top, left} = el.offset();

{
	top: top - document.body.scrollTop,
	left: left - document.body.scrollLeft
}
```

# Siblings

```js
[...el.parentNode.children].filter(child => child !== el);
```

# Trigger Custom

```js
el.dispatchEvent(new CustomEvent('my-event', {
	detail: {
		some: 'data'
	}
}));
```

# Trigger Native

`document.createEvent` has been deprecated in favour of constructors specific to the type of event.

# Array Each

```js
array.forEach((item, i) => {

});
```

# Deep Extend

```js
function deepExtend(target = {}, ...sources) {
	for (const source of sources) {
		if (!source) {
			continue;
		}

		for (const key in source) {
			if (!Object.prototype.hasOwnProperty.call(source, key)) {
				continue;
			}

			if (Array.isArray(source[key])) {
				target[key] = [...source[key]];
			} else if (typeof source[key] === 'object') {
				target[key] = deepExtend(target[key], source[key]);
			} else {
				target[key] = source[key];
			}
		}
	}

	return target;
}

deepExtend({}, objA, objB);
```

# Extend

```js
{
	...objA,
	...objB
}
```

# Map

```js
array.map((value, index) => {

});
```

# Parse HTML

```js
function parseHTML(input) {
	const element = document.implementation.createHTMLDocument();
	element.body.innerHTML = input;
	return element.body.children;
}

parseHTML(htmlString);
```

**Comment by ygoe** [comment]
@Richienb Your empty() replacement is not correct because it will always clone the node (losing event handlers?) that should be emptied (but otherwise not touched). The public version on this website is already good.

The replacement for $.extend() should probably be [Object.assign()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign). It seems to do exactly the same thing (if no deep inspection should be done).

**Comment by Richienb** [comment]
> The public version on this website is already good.

Just discovered this API that removes all children in one call: [`Element#replaceChildren`](https://developer.mozilla.org/en-US/docs/Web/API/Element/replaceChildren).

**Comment by ygoe** [comment]
One more:

```js
$(node).has(anotherNode).length > 0
```

becomes

```js
node.contains(anotherNode)
```

Note: While jQuery will not match the node itself, as per its documentation, the DOM method will.

https://developer.mozilla.org/en-US/docs/Web/API/Node/contains

**Comment by mscuthbert** [comment]
"Position Relative To Viewport" above assumes that el.offset() exists.  It doesn't.  But this is just as simple:

`
const {top, left} = el.getBoundingClientRect();
`

**Comment by nickserv** [comment]
Internet Explorer is officially dead, it would be nice to have the website show more modern examples by default, and skip recommending polyfills like fetch

**Comment by jarikmarwede** [comment]
How about still giving people different versions of compatibility but also options that are independent of browsers? For example by using JavaScript versions like `ES5` or `ES2017`.

**Comment by nickserv** [comment]
I don't think that's helpful because something being a JS feature could be part of ES2017 without being implemented in major browsers in ES2017. I'd argue that the average developer doesn't need to care about compatibility much any more.

**Comment by jarikmarwede** [comment]
I don't agree that the average developer doesn't need to care about compatibility. Internet Explorer shouldn't be a problem for most but for example from my experience Safari is always a little behind on supporting new features while Chrome releases features that are not even official standards yet. If there is an option like "I don't support Internet Explorer", what does it support? Maybe only official W3C Standards. But I don't know much about the different Standards stages. What's your opinion on this?

But I agree that JS versions could be a problem because of different browser support. Maybe linking features to [Can I Use](https://caniuse.com/) would be helpful.

**Comment by Abbondanzo** [comment]
Maybe there's another approach here that blends different browsers rather than different versions of each browser. IE is dead, sure, but what if I wanted to continue to support IE since many of my users access my website on extremely old hardware? (as of writing, that's still [~0.6% of Internet users](https://caniuse.com/usage-table))

Rather than a range slider for different IE versions, you can toggle on-and-off compatibility with each browser. Say you want to support Chrome, Firefox, and Safari, it'll show you the example alongside some hints as to the minimum versions of a browser where this implementation works. A link to caniuse.com can offer more in-depth compatibility for those who are even more curious.

The IE8/9/10 slider provides a level of granularity that I don't think is very useful for the vast majority of people who visit this site. Heck, [HubSpot stopped supporting IE11 August of last year](https://developers.hubspot.com/changelog/end-of-ie-support). I'd be happy to support carrying this website into the modern browser era since it's a very valuable resource

**Comment by nickserv** [comment]
Chrome, Firefox, and Safari already support everything this website uses and everything mentioned in this issue so far. As a result, we don't need a toggle for anything other than legacy/fringe browsers like Internet Explorer.

> @Abbondanzo The IE8/9/10 slider provides a level of granularity that I don't think is very useful for the vast majority of people who visit this site... I'd be happy to support carrying this website into the modern browser era since it's a very valuable resource

Do you mean you want to remove IE or have it behind a toggle?

**Comment by Abbondanzo** [comment]
> Do you mean you want to remove IE or have it behind a toggle?

Have it behind a toggle at the very most, or an indicator that certain implementations are _not_ supported on IE at the very least. No sense in throwing out the knowledge that has been accumulated here

**Comment by nickserv** [comment]
Here's what I would suggest:
1. Update code examples to support IE 11
2. Add examples using any modern JS features that are supported in Chrome, Firefox, and Safari
3. Change the IE slider to a checkbox or toggle that's disabled by default

**Comment by ygoe** [comment]
Sounds good. I mean, IE11 is deprecated. Earlier versions even more so.

**Comment by jarikmarwede** [comment]
But how would you handle situations where you need to support older versions of these browsers that don't support the new features. For example when you have to support older versions of Safari.

**Comment by nickserv** [comment]
Even very old versions of Safari still support these features.

**Comment by ygoe** [comment]
Why support a browser that isn't supported by its vendor? Are you suggesting to open security holes here? I've heard that jQuery 2.x supports old browsers, so you might want to use that if you're stuck in the past for some reason.

Or have I misunderstood something and this website aims to be a web museum? Then it should of course preserve all the old stuff it has. The older the better.

**Comment by nodsaibot** [comment]
Old versions and minorities are not important, Go forth!

**Comment by Abbondanzo** [comment]
Even MDN has [outright dropped IE](https://developer.mozilla.org/en-US/docs/Web/API/Event#browser_compatibility) out of their compatibility matrix. All the more reason to switch away from this. After #249 is merged, I think it's time to give this project a bit of an overhaul and start supporting new browsers out of the box.

After all, these are _code snippets_ and not production-ready, drop-in replacements designed to support every browser on earth. 

A couple of questions I want to answer:
- Is the minimum browser version a helpful distinction or does it add complexity when making decisions about what to support?
  - If yes, would it make more sense for us to simply attach minimum versions where these things were introduced? For example, fetch was added in Chrome 42, Edge 14, etc. 
  - If no or not really, what about linking directly to the compatibility matrices that MDN provides?
- How many browsers do we want to showcase? Chrome (& Edge, it's all the same), Firefox, Safari, IE make up a good showcase. As much as I love the idea of Opera, most people don't
- Are there more doors or wheels in the world?

**Comment by jarikmarwede** [comment]
> > Do you mean you want to remove IE or have it behind a toggle?
> 
> Have it behind a toggle at the very most, or an indicator that certain implementations are _not_ supported on IE at the very least. No sense in throwing out the knowledge that has been accumulated here

Is this still true then? Or have you changed your mind and would now suggest to remove all the old code instead of putting it behind a toggle?

**Comment by jarikmarwede** [comment]
> * Is the minimum browser version a helpful distinction or does it add complexity when making decisions about what to support?
>       
>   * If yes, would it make more sense for us to simply attach minimum versions where these things were introduced? For example, fetch was added in Chrome 42, Edge 14, etc.
>   * If no or not really, what about linking directly to the compatibility matrices that MDN provides?
> 
> * How many browsers do we want to showcase? Chrome (& Edge, it's all the same), Firefox, Safari, IE make up a good showcase. As much as I love the idea of Opera, most people don't

In my opinion adding code for modern browsers and either embedding compatibility data from somewhere like MDN or CanIUse, or just linking to their respective sites would be enough.

**Comment by Abbondanzo** [comment]
> Is this still true then? Or have you changed your mind and would now suggest to remove all the old code instead of putting it behind a toggle?

We can still keep the code around, no sense in getting rid of it, but it won't be visible by default

**Comment by nickserv** [comment]
> Is the minimum browser version a helpful distinction or does it add complexity when making decisions about what to support?

In my opinion, only for non-evergeen browsers like IE and Safari. Evergreen browsers update automatically so this isn't a concern for the majority of users.

Personally I think it would be fine to to just have everything support Chrome, Firefox, and Safari by default. If it turns out that some of the newer feature support in Safari is helpful, we could add a similar version slider like the one we had for IE (a lot of web devs believe Safari is the new IE anyway).

> If no or not really, what about linking directly to the compatibility matrices that MDN provides?

That's a good idea too, though I think it involves a bit more maintenance than what I suggested above.

> How many browsers do we want to showcase? Chrome (& Edge, it's all the same), Firefox, Safari, IE make up a good showcase. As much as I love the idea of Opera, most people don't

Opera also uses Chromium now.

**Comment by jarikmarwede** [comment]
> > If no or not really, what about linking directly to the compatibility matrices that MDN provides?
>
> That's a good idea too, though I think it involves a bit more maintenance than what I suggested above.

Can you explain why that would involve more maintenance? Wouldn't it just be one link that you would have to add while submitting the code for the first time?

**Comment by nickserv** [comment]
If I'm understanding @Abbondanzo's original comment, I believe it would be at least one link per feature, and we would have to figure out how to display them and keep them up to date.

**Comment by jarikmarwede** [comment]
Sorry, my comment was easy to misunderstand. I know that you would have to add a link for every comparison.
But why would you need to keep them up to date? I don't think CanIUse changes their links all that often. Or did I just misunderstand you?

**Comment by nickserv** [comment]
I mean if some new feature comes out, you'd have to make sure the new code has compatibility links, and that you were removing links for dead code.

**Comment by Abbondanzo** [comment]
> I mean if some new feature comes out, you'd have to make sure the new code has compatibility links, and that you were removing links for dead code.

In all fairness, that would probably mean we'd have to update our code snippets at the same time. There's maintenance overhead in any case

**Comment by mscuthbert** [comment]
I made a first PR at #271, which just adds one more option of "N/A" to the question of "What's the minimum version of IE that you support?"

My sense is that going forward, we could do something like:

1. Consolidate all the IE information into a single "IE" tag.  From work on #267 I can confirm that this tag would have `hide(versions['es6']); showFirst(versions['ie11'], versions['ie10'], ...etc...)` and it would display all the accumulated IE knowledge for anyone needing to support it still.

2. Rejigger the slider to say something like: "How many years old of a browser do you need to support?"  "9+ 7  5  3  1"  and start organizing the new files based on year that a feature was released; then add some JS to automagically show or hide versions based on the year released.

3. Or just turn it into a toggle with "Do you need to support IE?"  and concentrate on the actual comparisons / the actual jobs we get paid for.  :-)

**Comment by nickserv** [comment]
I think this can be closed since #271 has been merged.

We seem to be leaning toward option 3 for now (see #272).

**Comment by Abbondanzo** [comment]
Here's the direction we're going in before I close this issue out:

We're going to make this site _functional_, not to keep it around as a museum. That's what the Wayback Machine is for. As @mscuthbert has so eloquently put, it's the objective of this site to provide actual, real-world examples. As such, "I don't support IE" will be the default option going forward.
